File: apps/web/src/lib/format-money.ts
// Toman = Rial ÷ 10, display-scale only — see CLAUDE.md's Money & Date
// Conventions and the lifeos-domain skill. Every amount crosses the wire as
// a Rial minor-unit string (never a float, per MoneyAmountInput's contract);
// this file is the ONLY place that division happens, and it happens via
// BigInt, never parseFloat, to stay consistent with the project's
// never-float-money discipline all the way to the last mile.
const PERSIAN_DIGITS = ["۰", "۱", "۲", "۳", "۴", "۵", "۶", "۷", "۸", "۹"];

export function toPersianDigits(input: string): string {
  return input.replace(/[0-9]/g, (d) => PERSIAN_DIGITS[Number(d)]!);
}

function groupDigits(absValue: bigint): string {
  return absValue.toLocaleString("en-US"); // en-US grouping = plain "," every 3 digits, digit-set swapped separately for fa
}

// rial: a MoneyAmountInput-shaped string ("0" or a non-negative integer,
// no sign — the contract never emits negative amounts; INCOME/EXPENSE
// direction is a separate `type` field, not carried in the amount's sign).
export function formatTomanFromRial(rial: string, locale: "fa" | "en"): string {
  const toman = BigInt(rial) / 10n;
  const grouped = groupDigits(toman);
  return locale === "fa" ? toPersianDigits(grouped) : grouped;
}

// Un-grouped, un-localized Toman digits — used only to pre-fill an
// editable amount <Input> (where thousands separators would fight the
// user's cursor), as opposed to formatTomanFromRial's grouped/localized
// output for read-only display.
export function tomanRawFromRial(rial: string): string {
  return (BigInt(rial) / 10n).toString();
}

// Reverse direction: a user-typed Toman amount (plain digits, Persian or
// Latin) becomes the Rial minor-unit string the API expects.
export function parseTomanInputToRial(input: string): string {
  const latinDigits = input.replace(/[۰-۹]/g, (d) => String(PERSIAN_DIGITS.indexOf(d)));
  const digitsOnly = latinDigits.replace(/[^0-9]/g, "");
  if (digitsOnly === "") return "0";
  const toman = BigInt(digitsOnly);
  return (toman * 10n).toString();
}

File: mobile/lib/src/shared/format_money.dart
// Dart port of apps/web/src/lib/format-money.ts — same BigInt-only
// discipline (never parseDouble on a money string), same Toman = Rial ÷ 10
// display-scale conversion. See CLAUDE.md's Money & Date Conventions.
const _persianDigits = ['۰', '۱', '۲', '۳', '۴', '۵', '۶', '۷', '۸', '۹'];

String toPersianDigits(String input) {
  final buf = StringBuffer();
  for (final ch in input.split('')) {
    final d = int.tryParse(ch);
    buf.write(d == null ? ch : _persianDigits[d]);
  }
  return buf.toString();
}

String _groupDigits(BigInt absValue) {
  final s = absValue.toString();
  final buf = StringBuffer();
  for (var i = 0; i < s.length; i++) {
    if (i > 0 && (s.length - i) % 3 == 0) buf.write(',');
    buf.write(s[i]);
  }
  return buf.toString();
}

/// [rial] is a MoneyAmountInput/SignedMoneyAmount-shaped string from the
/// API (optionally "-" prefixed for signed fields like a wallet balance).
String formatTomanFromRial(String rial, {required bool fa}) {
  final value = BigInt.parse(rial);
  final toman = value ~/ BigInt.from(10);
  final sign = toman.isNegative ? '-' : '';
  final grouped = '$sign${_groupDigits(toman.abs())}';
  return fa ? toPersianDigits(grouped) : grouped;
}

/// Un-grouped Toman digits for pre-filling an editable amount field, where
/// thousands separators would fight the cursor.
String tomanRawFromRial(String rial) => (BigInt.parse(rial) ~/ BigInt.from(10)).toString();

/// Reverse direction: a user-typed Toman amount (Persian or Latin digits)
/// becomes the Rial minor-unit string the API expects.
String parseTomanInputToRial(String input) {
  final latin = input.split('').map((ch) {
    final i = _persianDigits.indexOf(ch);
    return i == -1 ? ch : i.toString();
  }).join();
  final digitsOnly = latin.replaceAll(RegExp(r'[^0-9]'), '');
  if (digitsOnly.isEmpty) return '0';
  return (BigInt.parse(digitsOnly) * BigInt.from(10)).toString();
}
